### PR TITLE
Add mlflow logging of metrics for each category

### PR DESCRIPTION
**Pull request by teodora-petkova**
Thank you really a lot for your branch and pull request  [#8261](https://github.com/ultralytics/yolov5/pull/8261) about mlflow logging!!

I have added a small option about logging the metrics for each category/class for every epoch as I did not see it available at the moment. I use it in my tests in order to analyze the progress of the metrics of each class and compare the separate metrics between runs, as well.

Here is a simple visualization for a few classes and their mAP@5.0 progression during all the epochs, that can be produced:
example

I am looking forward for your review! Thank you again!

### CODE REVIEW

**Comment by ElefHead** [comment]
This is great! I'm taking a vacation at the moment and will review this PR on satuday. Thanks so much!

**Comment by teodora-petkova** [comment]
Hi! Thank you! I have added a small change to take into account the case when the validation dataset does not contain examples from all classes. Also, I merged with the current state of the repository in order not to have any conflict with your base branch.
Is there something that I can help with? Or should I improve the pull request in some way?
Thank you a lot for you great work and help!

File: val.py [hunk 1]
```diff
@@ -276,6 +276,11 @@ def run(
         tp, fp, p, r, f1, ap, ap_class = ap_per_class(*stats, plot=plots, save_dir=save_dir, names=names)
         ap50, ap = ap[:, 0], ap.mean(1)  # AP@0.5, AP@0.5:0.95
         mp, mr, map50, map = p.mean(), r.mean(), ap50.mean(), ap.mean()
+
+        available_names = [names[c] for c in ap_class]
+                   
+        callbacks.run('on_val_end_compute_metrics', available_names, ap50, ap, f1, p, r)
```
Comment (ElefHead) [inline val.py:282]: I am gonna add this, but we might take it out if the repo maintainers don't want it -- wondering why no logging callbacks are present in the validation scripts.

File: utils/loggers/mlflow/mlflow_utils.py [hunk 2]
```diff
@@ -56,6 +56,7 @@ def setup(self, opt: Namespace) -> None:
         self.client = mlflow.tracking.MlflowClient()
         self.log_params(vars(opt))
         self.log_metrics(vars(opt), is_param=True)
+        self.current_epoch = {}
```
Comment (ElefHead) [inline utils/loggers/mlflow/mlflow_utils.py:59]: Should be set to 0? 
and might conflict on resumes - let me play around with this over the weekend, think there's different ways to handle resumes already

File: utils/loggers/__init__.py [hunk 3]
```diff
@@ -238,10 +240,20 @@ def on_val_end(self, nt, tp, fp, p, r, f1, ap, ap50, ap_class, confusion_matrix)
...
+        # log metrics for each class for every epoch
+        if self.mlflow:
+            for i in range(0, len(names)):
+                self.mlflow.log_metric(f"{names[i]}_mAP_0.5", ap50[i])
+                self.mlflow.log_metric(f"{names[i]}_mAP_0.5-0.95", ap[i])
+                self.mlflow.log_metric(f"{names[i]}_f1-score", f1[i])
+                self.mlflow.log_metric(f"{names[i]}_precision", p[i])
+                self.mlflow.log_metric(f"{names[i]}_recall", r[i])
+
     def on_fit_epoch_end(self, vals, epoch, best_fitness, fi):
```
Comment (ElefHead) [inline utils/loggers/__init__.py:257]: might need to handle `on_val_end` similar to how comet is handling it. Gonna implement that too.

**Review by ElefHead — CHANGES_REQUESTED** [review]

**Pull request closed without merging**